src/functions.jl

```diff
@@ -278,6 +278,109 @@ function Base.copy(f::ScalarQuadraticFunction)
     )
 end
 
+"""
+    ScalarNonlinearFunction(head::Symbol, args::Vector{Any})
+
+The scalar-valued nonlinear function `head(args...)`, represented as a symbolic
+expression tree, with the call operator `head` and ordered arguments in `args`.
+
+## `head`
+
+The `head::Symbol` must be an operator supported by the model.
+
+The default list of supported univariate operators is given by:
+
+ * [`Nonlinear.DEFAULT_UNIVARIATE_OPERATORS`](@ref)
+
+and the default list of supported multivariate operators is given by:
+
+ * [`Nonlinear.DEFAULT_MULTIVARIATE_OPERATORS`](@ref)
+
+Additional operators can be registered by setting a [`UserDefinedFunction`](@ref)
+attribute.
+
+See the full list of operators supported by a [`ModelLike`](@ref) by querying
+[`ListOfSupportedNonlinearOperators`](@ref).
+
+## `args`
+
+The vector `args` contains the arguments to the nonlinear function. If the
+operator is univariate, it must contain one element. Otherwise, it may contain
+multiple elements. Each element must be one of the following:
+
+ * A constant value of type `T<:Number`
```

Comment:
Should the element type of `args` be explicitly typed as a union of these types listed?

Reply:
@chriscoey and I had a chat about this.

The main argument for `Any` is simplicity. The downside is that the user might pass a disallowed argument like a vector of variables. When olvers like Ipopt which use `MOI.Nonlinear` encounter such an object, they will throw
https://github.com/jump-dev/MathOptInterface.jl/blob/acd7a0e72e4b736354d9dbe3b6c641ca3a541b05/src/Nonlinear/parse.jl#L33-L37

The other benefit is that if we are permissible in what types we accept, then a solver like MiniZinc could support vector arguments and work fine. But that's be non-standard syntax.

We could use the Union, but because the union is quite wide, it adds no performance benefit, and it'd make things clunkier to construct. I guess we could add a validator to the constructor?

Reply:
The method error principle should push us to type the array so that errors happen sooner rather than later.